bft/engine.go

```diff
@@ -0,0 +1,375 @@
+// Copyright (c) 2022 The VeChainThor developers
+
+// Distributed under the GNU Lesser General Public License v3.0 software license, see the accompanying
+// file LICENSE or <https://www.gnu.org/licenses/lgpl-3.0.html>
+package bft
+
+import (
+	"sort"
+	"sync/atomic"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/pkg/errors"
+	"github.com/vechain/thor/block"
+	"github.com/vechain/thor/builtin"
+	"github.com/vechain/thor/cache"
+	"github.com/vechain/thor/chain"
+	"github.com/vechain/thor/kv"
+	"github.com/vechain/thor/muxdb"
+	"github.com/vechain/thor/state"
+	"github.com/vechain/thor/thor"
+)
+
+const dataStoreName = "bft.engine"
+
+var finalizedKey = []byte("finalized")
+
+// BFTEngine tracks all votes of blocks, computes the finalized checkpoint.
+// Not thread-safe!
+type BFTEngine struct {
+	repo       *chain.Repository
+	data       kv.Store
+	voted      voted
+	master     thor.Address
+	stater     *state.Stater
+	forkConfig thor.ForkConfig
+	finalized  atomic.Value
+	caches     struct {
+		state   *lru.Cache
+		quality *lru.Cache
+		mbp     *lru.Cache
+		voteset *cache.PrioCache
+	}
+}
+
+// NewEngine creates a new bft engine.
+func NewEngine(repo *chain.Repository, mainDB *muxdb.MuxDB, forkConfig thor.ForkConfig, master thor.Address) (*BFTEngine, error) {
+	engine := BFTEngine{
+		repo:       repo,
+		data:       mainDB.NewStore(dataStoreName),
+		master:     master,
+		stater:     state.NewStater(mainDB),
+		forkConfig: forkConfig,
+	}
+
+	engine.caches.state, _ = lru.New(1024)
+	engine.caches.quality, _ = lru.New(1024)
+	engine.caches.mbp, _ = lru.New(8)
+	engine.caches.voteset = cache.NewPrioCache(16)
+
+	if val, err := engine.data.Get(finalizedKey); err != nil {
+		if !engine.data.IsNotFound(err) {
+			return nil, err
+		}
+		engine.finalized.Store(engine.repo.GenesisBlock().Header().ID())
+	} else {
+		engine.finalized.Store(thor.BytesToBytes32(val))
+	}
+
+	return &engine, nil
+}
+
+// Finalized returns the finalized checkpoint.
+func (engine *BFTEngine) Finalized() thor.Bytes32 {
+	return engine.finalized.Load().(thor.Bytes32)
+}
+
+// Accepts checks if the given block is on the same branch of finalized checkpoint.
+func (engine *BFTEngine) Accepts(parentID thor.Bytes32) error {
```

Comment:
how about to change the method signature to 
```
Accepts(blockID thor.Bytes32) (bool, error)
```
?

Then `errConflictWithFinalized` is not necessary  and can be  removed.